### PR TITLE
planner: respect `tidb_analyze_column_options ` when build analyze plan

### PR DESCRIPTION
**Pull request by hi-rustin**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/53567

This PR is based on https://github.com/pingcap/tidb/pull/54200.

Problem Summary:

When `tidb_analyze_column_options` is set to `PREDICATE`, we should only analyze predicate columns.

### What changed and how does it work?

- Checked `tidb_analyze_default_column_choice` when building the analyze plan.
- Only analyze indexes and the columns composing the indexes if there are no predicate columns.
- Added some tests to cover the analyze plan building code.
- Removed all use of `tidb_enable_column_tracking`.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/statistics/handle/usage/predicate_column_test.go [hunk 1]
```diff
@@ -54,3 +52,222 @@ func TestCleanupPredicateColumns(t *testing.T) {
...
+	tk.MustExec("create table t (a int, b int, c int)")
+	tk.MustExec("insert into t values (1, 1, 1), (2, 2, 2), (3, 3, 3)")
+	// Dump the statistics usage.
+	h := dom.StatsHandle()
+	err := h.DumpColStatsUsageToKV()
+	require.NoError(t, err)
+
+	// Analyze table.
+	tk.MustExec("analyze table t")
+	// FIXME: We should correct the job info or skip this kind of job.
```
Comment (hi-rustin) [inline pkg/statistics/handle/usage/predicate_column_test.go:269]: I will fix this problem in my next PR.

File: pkg/statistics/handle/usage/predicate_column_test.go [hunk 2]
```diff
@@ -54,3 +52,222 @@ func TestCleanupPredicateColumns(t *testing.T) {
...
+	// Analyze table.
+	tk.MustExec("analyze table t")
+	// Show Warnings.
+	tk.MustQuery("show warnings").Check(
+		testkit.Rows(
+			"Warning 1105 No predicate column has been collected yet for table test.t, so only indexes and the columns composing the indexes will be analyzed",
+			"Note 1105 Analyze use auto adjusted sample rate 1.000000 for table test.t, reason to use this rate is \"use min(1, 110000/10000) as the sample-rate=1\"",
+		),
+	)
+	// TODO: we should also include indexes in the job info.
```
Comment (hi-rustin) [inline pkg/statistics/handle/usage/predicate_column_test.go:245]: I will improve it in my next PR.

**Review by hi-rustin — COMMENTED** [review]
🔢 Self-check (PR reviewed by myself and ready for feedback.)

**Review by hi-rustin — COMMENTED** [review]
🔢 Self-check (PR reviewed by myself and ready for feedback.)

**Review by hi-rustin — COMMENTED** [review]
🔢 Self-check (PR reviewed by myself and ready for feedback.)

**Review by qw4990 — APPROVED** [review]

File: pkg/planner/core/planbuilder.go [hunk 3]
```diff
@@ -2017,22 +2021,40 @@ func (b *PlanBuilder) getFullAnalyzeColumnsInfo(
...
+				mustAnalyzedCols,
+				mustAllColumns,
+			)
+			if err != nil {
+				return nil, nil, err
+			}
+			return columns, nil, nil
+		default:
+			// Usually, this won't happen.
+			logutil.BgLogger().Warn("Unknown default column choice", zap.String("choice", columnOptions))
```
Comment (AilinKid) [inline pkg/planner/core/planbuilder.go:2042]: how about
```suggestion
			logutil.BgLogger().Warn("Unknown default column choice, analyze all columns", zap.String("choice", columnOptions))
```

File: pkg/statistics/handle/usage/predicate_column_test.go [hunk 4]
```diff
@@ -54,3 +52,245 @@ func TestCleanupPredicateColumns(t *testing.T) {
...
+}
+
+func TestAnalyzeWithNoPredicateColumnsAndNoIndexes(t *testing.T) {
+	store, dom := testkit.CreateMockStoreAndDomain(t)
+	tk := testkit.NewTestKit(t, store)
+
+	// Set tidb_analyze_column_options to PREDICATE.
+	tk.MustExec("set global tidb_analyze_column_options='PREDICATE'")
+
+	// Create table and select data without predicate.
```
Comment (AilinKid) [inline pkg/statistics/handle/usage/predicate_column_test.go:258]: didn't see select statement here？

File: pkg/statistics/handle/usage/predicate_column_test.go [hunk 5]
```diff
@@ -54,3 +52,245 @@ func TestCleanupPredicateColumns(t *testing.T) {
...
+	// Dump the statistics usage.
+	h := dom.StatsHandle()
+	err := h.DumpColStatsUsageToKV()
+	require.NoError(t, err)
+
+	// Analyze table.
+	tk.MustExec("analyze table t")
+	// FIXME: We should correct the job info or skip this kind of job.
+	tk.MustQuery("select table_name, job_info from mysql.analyze_jobs order by id desc limit 1").Check(
+		testkit.Rows("t analyze table columns  with 256 buckets, 100 topn, 1 samplerate"),
```
Comment (AilinKid) [inline pkg/statistics/handle/usage/predicate_column_test.go:271]: is that means all columns in L271's msg?

**Review by AilinKid — COMMENTED** [review]
Rest LGTM

File: pkg/statistics/handle/usage/predicate_column_test.go (hunk 4 above)
Comment (hi-rustin) [inline pkg/statistics/handle/usage/predicate_column_test.go:258]:
```suggestion
```

Comment (hi-rustin) [reply pkg/statistics/handle/usage/predicate_column_test.go]: Good catch! Deleted.

**Review by AilinKid — APPROVED** [review]

**Pull request merged**